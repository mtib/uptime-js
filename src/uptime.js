import axios from "axios";

export const checkUptimeAll = (servers) => {
    if (servers?.map) {
        return servers?.map?.(checkUptimeEntry);
    } else {
        return [];
    }
}

export const normalize = (server) => {
    return {
        url: server?.url,
        name: server?.name || server?.url,
        method: server?.method || 'GET',
        status: (() => {
            if (typeof server?.status === 'function') {
                return server.status;
            }
            if (typeof server?.status === 'object') {
                return (status) => server?.status.includes(status);
            }
            return (status) => status >= 200 && status < 300; 
        })(),
        payload: server?.payload,
        timeout: server?.timeout || 0,
    }
}

export const propagateDefault = (servers) => {
    const serverDefaults = {};
    return servers.map((server) => {
        if (server?._default) {
            Object.assign(serverDefaults, server._default);
        }
        Object.assign(server, serverDefaults);
        return normalize(server);
    });
}

export const checkUptimeEntry = (server) => {
    const start = Date.now();
    return axios({
        url: server?.url,
        method: server?.method,
        data: server?.payload,
        validateStatus: undefined,
        timeout: server?.timeout,
        responseType: 'arraybuffer',
    }).then((response) => {
        const end = Date.now();
        return { 
            success: (() => {
                if (typeof server?.status === 'function') {
                    return server.status;
                }
                if (typeof server?.status === 'object') {
                    return (status) => server?.status.includes(status);
                }
                return (status) => status >= 200 && status < 300; 
            })()(response.status), 
            status: response.status, 
            config: response.config,
            server: server,
            start: start,
            end: end,
            durationMs: end - start,
            response: response,
            size: response.data.length,
        };
    }).catch((error) => {
        return { success: false, status: error };
    });
}