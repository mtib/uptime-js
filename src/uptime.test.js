import { checkUptimeAll, checkUptimeEntry, normalize, propagateDefault } from "./uptime";
import jest from 'jest';

const minimalGithub = {
    url: 'https://github.com',
};

describe('successful requests', () => {
    test('minimal', async () => {
        await expect(checkUptimeEntry(minimalGithub)).resolves.toMatchObject({success: true});
    });

    test('minimal list', async () => {
        await expect(Promise.all(checkUptimeAll([minimalGithub, minimalGithub]))).resolves.toMatchObject([{success: true}, {success: true}]);
    });
});

describe('malformed single', () => {
    test('undefined', () => {
        expect(checkUptimeEntry()).toBeDefined();
        expect(checkUptimeEntry()).rejects;
    });
    test('null', () => {
        expect(checkUptimeEntry(null)).toBeDefined();
        expect(checkUptimeEntry(null)).rejects;
    });
})

describe('malformed list', () => {
    test('undefined', () => {
        expect(Promise.all(checkUptimeAll())).resolves.toBeDefined();
        expect(checkUptimeAll()).toMatchObject([]);
    });
    test('null', () => {
        expect(Promise.all(checkUptimeAll(null))).resolves.toBeDefined();
        expect(checkUptimeAll(null)).toMatchObject([]);
    });
})

describe('normalize', () => {
    test('minimal', () => {
        expect(normalize(minimalGithub)).toMatchSnapshot();
    });
    test('undefined', () => {
        expect(normalize()).toBeDefined();
    });
    test('null', () => {
        expect(normalize(null)).toBeDefined();
    });
})

describe('propagateDefaults', () => {
    test('propagateBeginning', () => {
        const servers = [
            {
                url: 'A',
                _default: {
                    timeout: 1000,
                },
            },
            {
                url: 'B',
            },
            {
                url: 'C',
            },
        ]
        const expectedTimeouts = [1000, 1000, 1000];
        expectedTimeouts.forEach((v, i) => {
            expect(propagateDefault(servers)?.[i]?.timeout).toBe(v);
        });
    })
    test('propagateMiddle', () => {
        const servers = [
            {
                url: 'A',
            },
            {
                url: 'B',
                _default: {
                    timeout: 1000,
                },
            },
            {
                url: 'C',
            },
        ]
        const expectedTimeouts = [0, 1000, 1000];
        expectedTimeouts.forEach((v, i) => {
            expect(propagateDefault(servers)?.[i]?.timeout).toBe(v);
        });
    })
    test('propagateTwice', () => {
        const servers = [
            {
                url: 'A',
                _default: {
                    timeout: 1000,
                },
            },
            {
                url: 'B',
            },
            {
                url: 'C',
                _default: {
                    timeout: 3000,
                },
            },
        ]
        const expectedTimeouts = [1000, 1000, 3000];
        expectedTimeouts.forEach((v, i) => {
            expect(propagateDefault(servers)?.[i]?.timeout).toBe(v);
        });
    })
})