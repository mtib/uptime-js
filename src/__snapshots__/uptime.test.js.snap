// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize minimal 1`] = `
Object {
  "method": "GET",
  "name": "https://github.com",
  "payload": undefined,
  "status": [Function],
  "timeout": 0,
  "url": "https://github.com",
}
`;
